components/posts/posts.php

```diff
@@ -253,6 +274,277 @@ function mm_posts_output_masonry_sizers() {
  * @param  object  $context  The global post object.
  * @param  array   $args     The instance args.
  */
+function mm_post_article( $post, $context, $args ) {
+
+	ob_start(); ?>
+
+	<article id="post-<?php the_ID( $post ); ?>" <?php post_class( 'mm-post' ); ?> itemscope itemtype="http://schema.org/BlogPosting" itemprop="blogPost" aria-label="Article">
+
+		<?php do_action( 'mm_posts_header', $post, $context, $args ); ?>
+
+		<?php do_action( 'mm_posts_content', $post, $context, $args ); ?>
+
+		<?php do_action( 'mm_posts_footer', $post, $context, $args ); ?>
+
+	</article>
+
+	<?php return ob_get_clean();
+}
+
+/**
+ * Output a taxonomy term filter.
+ *
+ * @since   1.0.0
+ *
+ * @param   WP_Query  $query    The query object.
+ * @param   WP_Post   $context  The global post object.
+ * @param   array     $args     The instance args.
+ *
+ * @return  string              The category filters HTML.
+ */
+function mm_posts_output_taxonomy_term_filter( $query, $context, $args ) {
+
+	$taxonomy = $args['taxonomy'] ? ( $args['taxonomy'] ) : 'category';
+
+	ob_start(); ?>
+
+	<?php if ( $args['filter_style'] == 'dropdown' ) : ?>
+
+		<div class="mm-posts-filter-wrapper">
+		<div class="mm-posts-filter" >
+		<?php wp_dropdown_categories(array(
+			'taxonomy'         => $taxonomy,
+			'orderby'          => 'name',
+	        'show_option_none' => 'Select',
+			'hide_empty'       => 1,
+	        'value_field'      => 'name',
+			'name'             => 'term-select',
+			'id'               => 'term_dropdown',
+		)); ?>
+
+		</div>
+		</div>
+
+	<?php else: ?>
+
+		<div class="mm-posts-filter-wrapper">
+		<ul class="mm-posts-filter" >
+		<li class="cat-item active"><a href="#" class="mm-posts-filter-all"><?php _e( 'All', 'mm-components' ); ?></a></li>
+		<?php wp_list_categories( array(
+			'title_li'     => '',
+			'hide_empty'   => 1,
+			'hierarchical' => false,
+			'taxonomy'     => $taxonomy,
+			)
+		); ?>
+		</ul>
+		</div>
+
+	<?php endif; ?>
+
+
+	<?php
+
+	return apply_filters( 'mm_posts_taxonomy_term_filter', ob_get_clean(), $query, $context, $args );
+}
+
+/**
+ * Echo the taxonomy term filter.
+ *
+ * @since   1.0.0
+ *
+ * @param   WP_Query  $query    The query object.
+ * @param   WP_Post   $context  The global post object.
+ * @param   array     $args     The instance args.
+ *
+ * @return  string              The category filters HTML.
+ */
+function mm_posts_taxonomy_term_filter( $query, $context, $args ) {
+	echo mm_posts_output_taxonomy_term_filter( $query, $context, $args );
+}
+
+/**
+ * Output taxonomy term filter data attributes.
+ *
+ * @since   1.0.0
+ *
+ * @param   WP_Query  $query    The query object.
+ * @param   WP_Post   $context  The global post object.
+ * @param   array     $args     The instance args.
+ *
+ * @return  string              The category filters HTML.
+ */
+function mm_posts_get_data_attributes( $query, $context, $args ) {
```

Comment:
This whole function can be DRYed up to become:

```
function mm_posts_get_data_attributes( $query, $context, $args ) {

	$data_atts = array();
	$data_args = array(
		'query_type',
		'post_ids',
		'post_type',
		'taxonomy',
		'term',
		'heading_level',
		'per_page',
		'pagination',
		'template',
		'show_featured_image',
		'featured_image_size',
		'show_post_info',
		'show_post_meta',
		'use_post_content',
		'filter_style',
		'link_title',
		'masonry',
		'current_page',
	);

	// Convert args to data-* attributes.
	foreach ( $data_args as $data_arg ) {
		if ( ! empty( $args[ $data_arg ] ) ) {
			$data_atts[] = 'data-' . str_replace( '_', '-', $data_arg ) . '="' . (string)$args[ $data_arg ] . '"';
		}
	}

	// Add special data attributes.
	$data_atts[] = 'data-total-pages="' . $query->max_num_pages . '"';
	$data_atts[] = 'data-total-posts="' . $query->found_posts . '"';

	// Convert to a string.
	$data_atts = esc_attr( implode( ' ', $data_atts ) );

	return apply_filters( 'mm_posts_data_attributes', $data_atts, $query, $context, $args );
}
```

There are a few things to note here. `$context` is the global post ID for the post/page that `mm_posts` is called on, but it's not really the "current" post ID, so if you wanted to track `$context->ID` as a data attribute here I would make it "data-global-post-id" and add it in the "Special attributes" section.

For the rest of these, I would always try to avoid string building and add things to arrays and `implode()` at the end, and `sanitize_text_field` is for values going into the DB. You want `esc_attr()` for any HTML attributes like these. For things like `$args['heading_level']` and `$args['per_page']` and `$args['current_page']` (the last of which you should probably add as a new `$arg` accepted by `mm_posts()`), since $args is already passed through `wp_parse_args( (array)$args, $defaults );` you already have default values for all the `$args`, so no need to specify any defaults anywhere else. It's always better to keep a single source of truth for default value setting, otherwise you'd have to touch multiple lines to update the default value for a single arg.

All of this said, since you really just care about getting all these values to pass to your JS, you might want to consider an alternate approach of just outputting a JS object as a variable in <script> tags so you don't have to parse any HTML attributes:

```
function mm_posts_get_js_data_var( $query, $context, $args ) {

	ob_start();

	?>
	<script type="text/javascript">
		var mmPostsData = {
			<?php

			// Output standard args.
			foreach ( $args as $key => $value ) {
				if ( ! empty( (string)$value ) ) {
					echo esc_attr( $key ) . ': "' . esc_attr( (string)$value ) . '"';
				}
			}

			// Output special extra args.
			echo 'total_pages: "' . esc_attr( $query->max_num_pages ) . '"';
			echo 'total_posts: "' . esc_attr( $query->found_posts ) . '"';

			?>
		};
	</script>
	<?php

	return apply_filters( 'mm_posts_js_data_var', ob_get_clean(), $query, $context, $args );
}
```
^I didn't test that, but it would be a much more direct way of getting at the data.